src/main/java/org/apache/skywalking/banyandb/v1/client/AbstractBulkWriteProcessor.java

```diff
@@ -44,33 +55,102 @@ public abstract class AbstractBulkWriteProcessor<REQ extends com.google.protobuf
      * @param flushInterval if given maxBulkSize is not reached in this period, the flush would be trigger
      *                      automatically. Unit is second.
      * @param concurrency   the number of concurrency would run for the flush max.
+     * @param timeout       network timeout threshold in seconds.
      */
-    protected AbstractBulkWriteProcessor(STUB stub, String processorName, int maxBulkSize, int flushInterval, int concurrency) {
-        super(processorName, maxBulkSize, flushInterval, concurrency);
+    protected AbstractBulkWriteProcessor(STUB stub,
+                                         String processorName,
+                                         int maxBulkSize,
+                                         int flushInterval,
+                                         int concurrency,
+                                         int timeout) {
         this.stub = stub;
+        this.maxBulkSize = maxBulkSize;
+        this.flushInterval = flushInterval;
+        this.timeout = timeout;
+        requests = new ArrayBlockingQueue<>(maxBulkSize + 1);
+        this.semaphore = new Semaphore(concurrency > 0 ? concurrency : 1);
+
+        scheduler = new ScheduledThreadPoolExecutor(1, r -> {
+            final Thread thread = new Thread(r);
+            thread.setName("ElasticSearch BulkProcessor");
+            return thread;
+        });
```

Comment:
This is the key to the new implementation. Semaphore is the concurrency lock to control race conditions between multiple threads flush and scheduler flush.